ons-spark/spark-concepts/shuffling.md

```diff
@@ -0,0 +1,755 @@
+
+In sparklyr we simply have `tbl_spark` objects rather than two APIs, e.g. to get the number of partitions use [`sdf_num_partitions()`](https://spark.rstudio.com/packages/sparklyr/latest/reference/sdf_num_partitions.html) on your sparklyr DataFrame.
+
+From a practical perspective when coding, do not worry too much about RDDs, the concept of a shuffle works exactly the same way on a DataFrame.
+
+### Examples
+
+To demonstrate shuffles we will run some Spark code, and see data being transferred between partitions.
+
+#### Example 1: One Shuffle
+
+Start with a small example to demonstrate the concept of shuffling and what actually happens to the data and partitions. First, create a Spark session. We are setting the number of partitions returned from a shuffled DataFrame to be just `2` with the `spark.sql.shuffle.partitions` option (the default is `200`).
+
+Note that in sparklyr with a local session `spark.sql.local.partitions` needs to be set; if running on a Spark cluster you can ignore this setting.
+````{tabs}
+```{code-tab} py
+import os
+from pyspark.sql import SparkSession, functions as F
+
+spark = (
+    SparkSession.builder.master("local[2]")
+    .appName("shuffling")
+    # Set number of partitions after a shuffle to 2
+    .config("spark.sql.shuffle.partitions", 2)
+    .getOrCreate()
+)
+```
+
+```{code-tab} r R
+
+library(sparklyr)
+library(dplyr)
+
+shuffle_config <- sparklyr::spark_config()
+# Set number of partitions after a shuffle to 2
+shuffle_config$spark.sql.shuffle.partitions <- 2
+# Set number of local partitions to 2
+shuffle_config$spark.sql.local.partitions <- 2
+
+sc <- sparklyr::spark_connect(
+    master = "local[2]",
+    app_name = "shuffling",
+    config = shuffle_config)
+
+```
+````
+As we will be using random numbers in this example we set the seed, so that the results can be replicated:
+````{tabs}
+```{code-tab} py
+seed_no = 999
+```
+
+```{code-tab} r R
+
+seed_no <- 999L
+
+```
+````
+For this example, create a DataFrame with one column, `id`, 20 rows and two partitions using [`spark.range()`](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.sql.SparkSession.range.html)/[`sdf_seq()`](https://spark.rstudio.com/packages/sparklyr/latest/reference/sdf_seq.html).
+
+We are caching the DataFrame after it is created with [`.cache()`](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.sql.DataFrame.cache.html)/[`tbl_cache()`](https://spark.rstudio.com/packages/sparklyr/latest/reference/tbl_cache.html), so that each action won't run the whole plan again from the start. This will make repeated actions faster and will not require recalculation of the random numbers. For more details on caching please see the article on Persisting.
+
+Now look at the DF:
+````{tabs}
+```{code-tab} py
+example_1 = spark.range(20, numPartitions=2)
+example_1.cache()
```

Comment:
I don't see the point in caching here. Does Spark do something differently if it isn't cached?

Reply:
The caching was a legacy from More Spark, to ensure that the random numbers stayed the same and to improve the speed when running it. Setting the seed resolves the random number consistency issue and we don't need to worry about speed, it's quick anyway.

Have removed all caching code.